### PR TITLE
feat: add support for new `TypeScript` versions

### LINKED ISSUE #23: New typescript syntax is not supported
I noticed that some typescript features are not supported. Example `keyof`

An invalid golden files is being generated

Example instead of the below;
```ts
/** @stable */
export declare type SimpleChanges<T = any> = {
    [P in keyof T]?: SimpleChange<T[P]>;
};
/** @stable */
export declare const SkipSelf: SkipSelfDecorator;
```
it is generating the below in public_api_guard/core/core.d.ts
```ts
/** @stable */
export declare type SimpleChanges<T = any> = {
    [P in keyof

/** @stable */
export declare const SkipSelf: SkipSelfDecorator;
```

https://github.com/angular/angular/pull/22103#discussion_r167059184

### PR DESCRIPTION
**Pull request by alan-agius4**
There are several changes:

- Update to `TypeScript 2.6+`
- Remove `dts` in favor of `@types`
- Add package-lock
- Update several dependencies including `chalk`, `merge2`, `gulp-typescript`
- Add `TypeScript` as a `peerDependency`
- Add a new fixture to test `keyOf`

Closes: #23

### CODE REVIEW

**Comment by alan-agius4** [comment]
@mhevery can you have a look please?

**Comment by alan-agius4** [comment]
@alexeagle can you have a look please?

File: lib/serializer.ts [hunk 1]
```diff
@@ -27,16 +27,22 @@ export interface SerializationOptions {
    * Warns or errors if stability annotations are missing on an export.
    * Supports experimental, stable and deprecated.
    */
-  onStabilityMissing?: string;  // 'warn' | 'error' | 'none'
+  onStabilityMissing?: DiagnosticType;
 }
 
+export enum DiagnosticType {
```
Comment (alexeagle) [inline lib/serializer.ts]: why an enum, if you just want the three string values it should be
export type DiagnosticType = 'warn'|'error'|'none';
(and I'd name it DiagnosticSeverity)

**Review by alexeagle — COMMENTED** [review]
Thanks, this looks awesome!

I think we should test it against angular/angular too - if you don't have time for that, I can do it.

**Comment by alan-agius4** [comment]
If you’d can it will be awesome if you test it with Angular.

Ill do the changes for the enum tomorrow, as I dont have access to my
machine today. If feel that it needs to done today. Feel free to update my
PR :)

File: lib/serializer.ts (hunk 1 above)
Comment (alan-agius4) [reply lib/serializer.ts]: Ill do the changes for the enum tomorrow, as I dont have access to my
machine today. If feel that it needs to done today. Feel free to update my
PR :)

**Comment by alan-agius4** [comment]
It would be great if you can test it with angular as at the moment I don’t have my mac handly until the weekend and on my windows I always have problems to run gulp update:api inside of angular

Thanks

File: lib/serializer.ts (hunk 1 above)
Comment (alan-agius4) [reply lib/serializer.ts]: hey managed to do the requested changed today. I just pushed it.

Comment (alan-agius4) [reply lib/serializer.ts]: Just a quick question, when would you consider using string enums over string union types?

**Comment by alexeagle** [comment]
I tried this out in the angular repo and it produced a ton of deltas. I tried to turn them into a PR against angular/angular but something is broken with the `gulp public-api:update` script in angular, it just hangs (not sure if that is related to this change)
https://pastebin.com/xBEPaaz5 is a sample of the delta.

Is it expected that this PR would change the public API golden files so much?

**Comment by alan-agius4** [comment]
Thanks for taking the time into looking into this.

No, those changes are not expected. To be honest they are mainly ordering related and are based on the following; https://github.com/angular/ts-api-guardian/blob/master/lib/serializer.ts#L268

I'll have a look at them and see what's the issue.

**Comment by alan-agius4** [comment]
Hey, I pushed an update that should fix the ordering. I also added a Unit Test to cover this. https://github.com/angular/ts-api-guardian/pull/25/commits/e088033ef5d794d6d6a55372400f24f1c88699b9 Can you try to have a another look?

Thanks a lot.

**Comment by alan-agius4** [comment]
I’ve tested it with Angular repo, and it’s looking good. i created a PR and tagged you.

**Comment by alan-agius4** [comment]
Hey @alexeagle, I’ve tested it with Angular repo, and it’s looking good. i created a PR and tagged you.

**Comment by alexeagle** [comment]
Awesome, thank you for making this mergable. FYI after Angular v6 is released we plan to move this repo into angular/angular, which will make this easier since the Angular API update will be part of the CI for changing ts-api-guardian

**Comment by alan-agius4** [comment]
Thanks for the info 😁, it does make sense indeed to move it to angular/angular and maybe it will get maintaned a bit more. Cause I don’t think many people know about this.

**Pull request merged**